### PR TITLE
Update moderation commands

### LINKED ISSUE #55: As a moderator, i Need to be DM certain Users or everyone in a Server
Commands
-----------
`dm <list of users> <message>` : This DMs all the specified users with given message

`dmall <message>` : This DMs all the users in my server with The given Message

Permissions required for this : Since it's only for moderators, the permissions can be `kick_members=True` or `ban_members=True`

### PR DESCRIPTION
**Pull request by janaSunrise**
Closes #55 

Added shift command : Shifts messages from one channel, to a specified channel
Update Clear command : specify a user to, clear only his message, or clear everything
Added Dm command : Specify a List of users and a message, to DM them, in a server
Added DMall Command : Specify a message to Send to all the users in a server

### CODE REVIEW

File: cogs/moderation.py [hunk 1]
```diff
@@ -226,11 +226,13 @@ async def multiban(self, ctx: Context, members: Greedy[UserID], *, reason: str =
 
         total_members = len(members)
         if total_members == 0:
-            return await ctx.send("No members to ban.")
+            await ctx.send("No members to ban.")
+            return
```
Comment (Faholan) [inline cogs/moderation.py:230]: Why ? It was fine as it was

Comment (janaSunrise) [reply cogs/moderation.py:230]: the return type is None

File: cogs/moderation.py [hunk 2]
```diff
@@ -293,6 +298,54 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+        async for message in ctx.message.channel.history(limit=count):
+            embed = Embed(description=message.content, color=Color.green())
+            embed.set_author(name=message.author.name, icon_url=message.author.avatar_url)
+            embed.timestamp = message.created_at
+            messages.append(embed)
+
+            if not copy:
+                await message.delete()
+
+        await target.send(f'Message source : {ctx.message.channel.mention}.')
```
Comment (Faholan) [inline cogs/moderation.py:315]:
```suggestion
        await target.send(f'Message source : {ctx.channel.mention}.')
```

File: cogs/moderation.py [hunk 3]
```diff
@@ -293,6 +298,54 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+        embed = Embed(
+            title="Notice!",
+            description=message,
+            color=Color.dark_magenta()
+        )
+        embed.set_footer(text=f"From {ctx.guild.name}", icon_url=ctx.guild.icon_url)
+
+        for member in ctx.guild.members:
+            with suppress(Forbidden):
+                await member.send(embed=embed)
```
Comment (Faholan) [inline cogs/moderation.py:347]: A command to opt out from this would be welcome

**Review by Faholan — DISMISSED** [review]
To add : command to opt out from DMs

File: cogs/moderation.py (hunk 2 above)
Comment (ItsDrike) [reply cogs/moderation.py:315]: You should also move the `:` there shouldn't be a space before it:
```suggestion
        await target.send(f'Message source: {ctx.message.channel.mention}.')
```

File: cogs/moderation.py (hunk 3 above)
Comment (ItsDrike) [reply cogs/moderation.py:347]: Technically you can opt-out, you can disable direct messages for any given server, it's a discord built-in feature
image

File: cogs/moderation.py [hunk 4]
```diff
@@ -293,6 +298,54 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+            if not copy:
+                await message.delete()
+
+        await target.send(f'Message source : {ctx.message.channel.mention}.')
+
+        for embed in reversed(messages):
+            await target.send(embed=embed)
+
+    @command()
+    @has_permissions(kick_members=True)
```
Comment (ItsDrike) [inline cogs/moderation.py:321]: This is unacceptable, as I pointed out in https://github.com/The-Codin-Hole/HotWired-Bot/issues/55#issuecomment-649825269 we can't just use kick permissions check

File: cogs/moderation.py [hunk 5]
```diff
@@ -293,6 +298,54 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+            description=message,
+            color=Color.dark_magenta()
+        )
+        embed.set_footer(text=f"From {ctx.guild.name}", icon_url=ctx.guild.icon_url)
+
+        for member in members:
+            await member.send(embed=embed)
+
+    @command()
+    @has_permissions(kick_members=True)
```
Comment (ItsDrike) [inline cogs/moderation.py:335]: Same as above (Administrator permission)

**Review by ItsDrike — CHANGES_REQUESTED** [review]

File: cogs/moderation.py [hunk 6]
```diff
@@ -276,9 +278,12 @@ async def unban(self, ctx: Context, *, user: UserID) -> None:
 
     @command()
     @has_permissions(manage_messages=True)
-    async def clear(self, ctx: Context, amount: int) -> None:
+    async def clear(self, ctx: Context, amount: int, target: User = None) -> None:
```
Comment (ItsDrike) [inline cogs/moderation.py:281]: Might be a good idea to match UserID instead (The custom converter)

File: cogs/moderation.py [hunk 7]
```diff
@@ -293,6 +298,96 @@ async def clear(self, ctx: Context, amount: int) -> None:
         await asyncio.sleep(2.5)
         await message.delete()
 
+    @command()
+    @has_permissions(manage_messages=True)
+    async def shift(self, ctx, count: int, target: TextChannel, copy: bool = False) -> None:
+        """Copy or Move specified messages amount to specified channel"""
+        messages = []
+        async for message in ctx.message.channel.history(limit=count):
```
Comment (ItsDrike) [inline cogs/moderation.py:306]:
```suggestion
        async for message in ctx.channel.history(limit=count):
```
I don't think the message is necessary here.

File: cogs/moderation.py [hunk 8]
```diff
@@ -293,6 +298,96 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+        """Dm a List of Specified User from Your Guild."""
+        embed = Embed(
+            title="Notice!",
+            description=message,
+            color=Color.dark_magenta()
+        )
+        embed.set_footer(text=f"From {ctx.guild.name}", icon_url=ctx.guild.icon_url)
+
+        for member in members:
+            await member.send(embed=embed)
```
Comment (ItsDrike) [inline cogs/moderation.py:332]: It would be nice if you could also pass in roles and DM everyone in that role like that.
Do note that doing that would mean that the user will need administrator permissions, not just manage messages

File: cogs/moderation.py [hunk 9]
```diff
@@ -293,6 +298,96 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+            with suppress(Forbidden):
+                await member.send(embed=embed)
+
+    @command()
+    @has_permissions(manage_roles=True)
+    async def lock(self, ctx, channels: Greedy[TextChannel] = None, reason: str = 'Not Specified') -> None:
+        """Disable @everyone's permission to send message on given channel or current channel if not specified."""
+        if channels is None:
+            channels = [ctx.channel]
+        for c in channels:
```
Comment (ItsDrike) [inline cogs/moderation.py:355]: use `channel` instead of `c`, it's much more descriptive

File: cogs/moderation.py [hunk 10]
```diff
@@ -293,6 +298,96 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+            color=Color.dark_magenta()
+        )
+        embed.set_footer(text=f"From {ctx.guild.name}", icon_url=ctx.guild.icon_url)
+
+        for member in ctx.guild.members:
+            with suppress(Forbidden):
+                await member.send(embed=embed)
+
+    @command()
+    @has_permissions(manage_roles=True)
```
Comment (ItsDrike) [inline cogs/moderation.py:350]: We shouldn't check for `manage_roles` for this, but instead, manage permissions for that given channel together with manage messages, that makes much more sense, why would a user need to be able to manage roles in order to lock a channel.

Note: This also applies for `unlock`.

File: cogs/moderation.py [hunk 11]
```diff
@@ -293,6 +298,96 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+        if not channels == [ctx.channel]:
+            await ctx.send(f"Locked down {len(channels)} channel{'s' if len(channels) > 1 else ''}.")
+
+    @command()
+    @has_permissions(manage_roles=True)
+    async def unlock(self, ctx, channels: Greedy[TextChannel] = None, reason: str = 'Not specified') -> None:
+        """Reset @everyone's permission to send message on given channel or current channel if not specified."""
+        if channels is None:
+            channels = [ctx.channel]
+        for c in channels:
```
Comment (ItsDrike) [inline cogs/moderation.py:368]: Change `c` to `channel` (as described above)

File: cogs/moderation.py [hunk 12]
```diff
@@ -293,6 +298,96 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+            await c.set_permissions(c.guild.default_role, send_messages=None,
+                                    reason=f"Reason: {reason} | Requested by {ctx.author}.")
+        await ctx.send(f"Unlocked {len(channels)} channel{'s' if len(channels) > 1 else ''}.")
+
+    @command()
+    @has_permissions(manage_channels=True)
+    async def slowmode(self, ctx, channels: Greedy[TextChannel] = None, seconds: int = 10, reason: str = "Not specified") -> None:
+        """Set channel's slowmode delay, default = 10s."""
+        if not (0 <= seconds <= 21600):
+            await ctx.send(":x: The duration is extreme or very low.")
```
Comment (ItsDrike) [inline cogs/moderation.py:378]:
```suggestion
            await ctx.send(":x: Duration is out of bounds (0-21600 seconds)")
```

File: cogs/moderation.py [hunk 13]
```diff
@@ -293,6 +298,96 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+            return
+
+        if channels is None:
+            channels = ctx.channel
```
Comment (ItsDrike) [inline cogs/moderation.py:382]: The variable name is `channels` it should be referencing to multiple channels, so it should be a list
```suggestion
            channels = [ctx.channel]
```

File: cogs/moderation.py [hunk 14]
```diff
@@ -293,6 +298,96 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+
+        for c in channels:
```
Comment (ItsDrike) [inline cogs/moderation.py:384]: Once again, undescriptive name, change `c` to `channel`

**Review by ItsDrike — CHANGES_REQUESTED** [review]
Some minor adjustments

File: cogs/moderation.py (hunk 10 above)
Comment (janaSunrise) [reply cogs/moderation.py:350]: uhm, its not possible.

File: assets/checks.py [hunk 15]
```diff
@@ -1,12 +1,23 @@
...
 
+import typing as t
+from cogs.utils import constants
+
 
 def owner(ctx: Context) -> bool:
     return ctx.author.id in owner_ids
 
 
+async def is_sudo(ctx: Context) -> t.Union[bool, None]:
```
Comment (e-xe) [inline assets/checks.py:13]:
```suggestion
async def is_dev(ctx: Context) -> t.Union[bool, None]:
```

File: assets/checks.py [hunk 16]
```diff
@@ -1,12 +1,23 @@
...
+    if ctx.author.id in constants.devs:
+        return True
+    else:
+        embed = Embed(description="Make your Own sandwich Mortal!.", color=Color.red())
+        await ctx.send(embed=embed)
```
Comment (e-xe) [inline assets/checks.py:18]:
```suggestion
    else:
        return True
```
don't react if the user is not a dev

File: cogs/moderation.py [hunk 17]
```diff
@@ -293,6 +298,104 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+        embed = Embed(
+            title="Notice!",
+            description=message,
+            color=Color.dark_magenta()
+        )
+        embed.set_footer(text=f"From {ctx.guild.name}", icon_url=ctx.guild.icon_url)
+
+        for member in ctx.guild.members:
+            with suppress(Forbidden):
+                await member.send(embed=embed)
```
Comment (e-xe) [inline cogs/moderation.py:355]: Add a title parameter

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by ItsDrike** [comment]
This PR requires an update from the author ( @janaSunrise ) which bumps it up to respect #62 which introduced certain changes to `moderation.py` but also overall update from `master`, the author needs to either revisit this PR or close it.

File: bot/cogs/moderation.py [hunk 18]
```diff
@@ -193,6 +199,104 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+                return
+            else:
+                await member.send(embed=embed)
+
+    @command()
+    @has_permissions(administrator=True)
+    async def dmall(self, ctx: Context, *, message: str) -> None:
+        """Dm all Users from Your Guild."""
+        embed = Embed(
+            title="Notice!",
```
Comment (e-xe) [inline bot/cogs/moderation.py:248]: Don't hardcore titles

**Review by ghost — CHANGES_REQUESTED** [review]

File: bot/cogs/moderation.py (hunk 18 above)
Comment (janaSunrise) [reply bot/cogs/moderation.py:248]: we can't use dynamic titles

Comment (janaSunrise) [reply bot/cogs/moderation.py:248]: if we do this : `async def dmall(self, ctx: Context, *, title: str, *, message: str) -> None:`
which is gonna be title, and which one is message?

Comment (e-xe) [reply bot/cogs/moderation.py:248]: Something like this:
```
>>dmall title="Important mail" description="Desripto"

Comment (janaSunrise) [reply bot/cogs/moderation.py:248]: uh, using kwargs in bot commands?
not possible

**Review by ActuallyDaneel — CHANGES_REQUESTED** [review]
Please change the prefix back to ">>"
Also there are several merge conflicts

File: bot/cogs/events.py [hunk 19]
```diff
@@ -117,9 +120,29 @@ async def on_guild_join(self, guild: Guild) -> None:
...
+
+                Text channels: **{len(guild.text_channels)}**
+                Voice channels: **{len(guild.voice_channels)}**
+                AFK timeout: **{round(guild.afk_timeout / 60)}m | AFK channel: {guild.afk_channel}**
+
+                Created at: **{datetime.datetime.strftime(guild.created_at, "%A %d %B %Y at %H:%M")}**
+                File Size: **{round(guild.filesize_limit / 1048576)} MB**
+                Bitrate: **{round(guild.bitrate_limit / 1000)} kbps**
+                """
+            ),
```
Comment (e-xe) [inline bot/cogs/events.py:141]: Why would we need this? The current embed is better imo

**Review by ghost — CHANGES_REQUESTED** [review]

File: bot/cogs/events.py (hunk 19 above)
Comment (janaSunrise) [reply bot/cogs/events.py:141]: we need it, later if the bot is abused, this can be used to find the server, and report it

Comment (therealdaneel) [reply bot/cogs/events.py:141]: It's definitely kinda pointless but what's the harm in keeping it @e-xe ?

Comment (e-xe) [reply bot/cogs/events.py:141]: So if someone abuses the bot from a server, will we go to the log channel, looking for the server info? Or can't we just send an email to discord with the server id?
And I see some unneeded information in there such as bit-rate, file size etc. All of them, actually, besides the server id and name and what server we are on.
```suggestion
        log_embed = Embed(
            title=f"Bot is invited to a server!",
            color=Color.green(),
            description=textwrap.dedent(
                f"""
                **We've reached our {len(self.bot.guilds)}th server!** :champagne_glass:
                Server name: {guild.name} | Id: {guild.id}
                """
            ),
```

File: bot/cogs/events.py [hunk 20]
```diff
@@ -117,9 +120,23 @@ async def on_guild_join(self, guild: Guild) -> None:
...
+                f"""
+                **We've reached our {len(self.bot.guilds)}th server!** :champagne_glass:
+                Guild Owner: **{guild.owner}** | **<@!{guild.owner_id}>**
+
+                Member Count: **{guild.member_count}**
+
+                Created at: **{datetime.datetime.strftime(guild.created_at, "%A %d %B %Y at %H:%M")}**
+                """
+            ),
         )
```
Comment (e-xe) [inline bot/cogs/events.py:136]:
```suggestion
        log_embed = Embed(
            title=f"The bot has been added to {guild.name}",
            color=Color.green(),
            description=textwrap.dedent(
                f"""
                **We've reached our {len(self.bot.guilds)}th server!** :champagne_glass:
                Guild id: **{guild.id}**
                Guild Owner: **{guild.owner}** | **<@!{guild.owner_id}>**

                Member Count: **{guild.member_count}**

                Created at: **{datetime.datetime.strftime(guild.created_at, "%A %d %B %Y at %H:%M")}**
                """
            ),
        )
```

File: bot/cogs/events.py [hunk 21]
```diff
@@ -117,9 +120,23 @@ async def on_guild_join(self, guild: Guild) -> None:
...
+            description=textwrap.dedent(
+                f"""
+                **We've reached our {len(self.bot.guilds)}th server!** :champagne_glass:
+                Guild Owner: **{guild.owner}** | **<@!{guild.owner_id}>**
+
+                Member Count: **{guild.member_count}**
+
+                Created at: **{datetime.datetime.strftime(guild.created_at, "%A %d %B %Y at %H:%M")}**
+                """
+            ),
```
Comment (e-xe) [inline bot/cogs/events.py:135]:
```suggestion
        log_embed = Embed(
            title=f"The bot has been added to {guild.name}",
            color=Color.green(),
            description=textwrap.dedent(
                f"""
                **We've reached our {len(self.bot.guilds)}th server!** :champagne_glass:
                Guild id: **{guild.id}**
                Guild Owner: **{guild.owner}** | **<@!{guild.owner_id}>**

                Member Count: **{guild.member_count}**

                Created at: **{datetime.datetime.strftime(guild.created_at, "%A %d %B %Y at %H:%M")}**
                """
            ),
```

File: bot/cogs/events.py [hunk 22]
```diff
@@ -117,9 +120,23 @@ async def on_guild_join(self, guild: Guild) -> None:
...
             pass
 
-        await logchannel.send(
-            f"The bot has been added to **{guild.name}** , " f"We've reached our **{len(self.bot.guilds)}th** server! :champagne_glass: "
+        log_embed = Embed(
+            title=f"The bot has been added to {guild.name}",
+            color=Color.green(),
+            description=textwrap.dedent(
+                f"""
+                **We've reached our {len(self.bot.guilds)}th server!** :champagne_glass:
```
Comment (e-xe) [inline bot/cogs/events.py:128]:
```suggestion
                **We've reached our {len(self.bot.guilds)}th server!** :champagne_glass:
                Guild Id: **{guild.id}**
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: bot/cogs/sudo.py [hunk 23]
```diff
@@ -44,72 +52,64 @@ def get_uptime(self) -> str:
...
-            await ctx.send(embed=embed)
-
     @group(hidden=True)
     async def sudo(self, ctx: Context) -> None:
         """Administrative information."""
         pass
 
     @sudo.command()
-    @check(is_owner)
+    @check(is_sudo)
```
Comment (e-xe) [inline bot/cogs/sudo.py:61]: This name doesn't make any sense
Sudo stands for **Su**peruser**Do**
And this command now stands for `is_superuserdo`

**Review by ghost — CHANGES_REQUESTED** [review]

File: bot/cogs/moderation.py [hunk 24]
```diff
@@ -181,12 +199,108 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+
+    @command()
+    @has_permissions(manage_channels=True)
+    async def unlock(self, ctx, channels: Greedy[TextChannel] = None, reason: str = 'Not specified') -> None:
+        """Reset @everyone's permission to send message on given channel or current channel if not specified."""
+        if channels is None:
+            channels = [ctx.channel]
+        for channel in channels:
+            await channel.set_permissions(channel.guild.default_role, send_messages=None,
+                                          reason=f"Reason: {reason} | Requested by {ctx.author}.")
```
Comment (ItsDrike) [inline bot/cogs/moderation.py:283]:
```suggestion
            await channel.set_permissions(
                channel.guild.default_role, 
                send_messages=None,
                reason=f"Reason: {reason} | Requested by {ctx.author}."
            )
```

File: bot/cogs/moderation.py [hunk 25]
```diff
@@ -181,12 +199,108 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+        await ctx.send(f"Unlocked {len(channels)} channel{'s' if len(channels) > 1 else ''}.")
+
+    @command()
+    @has_permissions(manage_channels=True)
```
Comment (ItsDrike) [inline bot/cogs/moderation.py:287]: There might be channels where users with global `manage_channels` permission have this permission explicitly disabled (or vice-versa: users with this permission without having it as global), we should check this permission for every channel they try to use this on instead of checking globally.

File: bot/cogs/sudo.py [hunk 26]
```diff
@@ -32,6 +40,13 @@ def __init__(self, bot: Bot) -> None:
         self.bot = bot
         self.start_time = datetime.datetime.utcnow()
 
+    def cog_check(self, ctx: Context) -> t.Optional[bool]:
+        if ctx.author.id in devs:
+            return True
+        else:
+            embed = Embed(description="Make your Own sandwich Mortal!.", color=Color.red())
+            await ctx.send(embed=embed)
```
Comment (ItsDrike) [inline bot/cogs/sudo.py:48]: We don't need to use `else` since when the first if passes it will go directly to `return True` ending the function, only if it doesn't go there it would always go to `else` which means we might as well use flatter code (which is better according to the zen of python) instead.

We should also return `False` in case the check isn't passed
```suggestion
        if ctx.author.id in devs:
            return True
 
        embed = Embed(description="Make your Own sandwich Mortal!.", color=Color.red())
        await ctx.send(embed=embed)
        return False
```

File: bot/cogs/sudo.py [hunk 27]
```diff
@@ -44,72 +59,59 @@ def get_uptime(self) -> str:
...
     @group(hidden=True)
     async def sudo(self, ctx: Context) -> None:
         """Administrative information."""
         pass
 
     @sudo.command()
-    @check(is_owner)
     async def shutoff(self, ctx: Context) -> None:
-        if ctx.author.id in config.devs:
+        if ctx.author.id in devs:
```
Comment (ItsDrike) [inline bot/cogs/sudo.py:69]: This is pointless due to the `cog_check`

**Review by ItsDrike — CHANGES_REQUESTED** [review]

File: bot/cogs/moderation.py [hunk 28]
```diff
@@ -181,12 +198,137 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
         )
         message = await ctx.send(ctx.author.mention, embed=embed)
         await asyncio.sleep(2.5)
         await message.delete()
 
+    @command()
+    @has_permissions(manage_messages=True)
+    async def shift(self, ctx, count: int, target: TextChannel, copy: bool = False) -> None:
+        """Copy or Move specified messages amount to specified channel"""
+        messages = []
```
Comment (therealdaneel) [inline bot/cogs/moderation.py:211]:
```suggestion
        """Copy or Move specified messages amount to specified channel"""
        if not (5 <= count <= 100):
            await ctx.send("Amount of messages shifted must be greater than 0 and smaller than 100")
        messages = []
```
I suggest this because otherwise this command may end up being rather spammy due to rate-limiting

File: bot/cogs/moderation.py [hunk 29]
```diff
@@ -181,12 +198,137 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+            embed.timestamp = message.created_at
+            messages.append(embed)
+
+            if not copy:
+                await message.delete()
+
+        await target.send(f'Message source : {ctx.channel.mention}.')
+
+        for embed in reversed(messages):
+            await target.send(embed=embed)
```
Comment (therealdaneel) [inline bot/cogs/moderation.py:224]:
```suggestion
        for embed in reversed(messages):
            await target.send(embed=embed)
            await asyncio.sleep(0.5)
```
again, to prevent or at least delay rate-limit staggering

File: bot/cogs/moderation.py [hunk 30]
```diff
@@ -2,26 +2,38 @@
 import textwrap
 import typing as t
 from collections import Counter
```
Comment (therealdaneel) [inline bot/cogs/moderation.py:4]:
```suggestion
import textwrap
import typing as t
import asyncio
from collections import Counter
```

**Review by ActuallyDaneel — CHANGES_REQUESTED** [review]
basically an attempt to reduce the impact of discords rate-limiting for the `shift` command

File: bot/config.py [hunk 31]
```diff
@@ -35,7 +35,7 @@
 SUPPORT_SERVER = "https://discord.gg/7e9zKFr"
 
 
-COMMAND_PREFIX = os.getenv("COMMAND_PREFIX", ">>")
+COMMAND_PREFIX = os.getenv("COMMAND_PREFIX", ";")
```
Comment (therealdaneel) [inline bot/config.py:38]:
```suggestion
COMMAND_PREFIX = os.getenv("COMMAND_PREFIX", ">>")
```
why did you change the prefix again?

**Review by ActuallyDaneel — CHANGES_REQUESTED** [review]

File: bot/cogs/moderation.py [hunk 32]
```diff
@@ -120,17 +132,19 @@ async def multiban(self, ctx: Context, members: Greedy[ProcessedMember], *, reas
...
         if total_members == 0:
-            return await ctx.send("No members to ban.")
+            await ctx.send("No members to ban.")
+            return
 
         confirm = await ctx.prompt(f"This will ban **{Plural(total_members):member}**. Are you sure?", reacquire=False)
         if not confirm:
-            return await ctx.send("Aborting.")
+            await ctx.send("Aborting.")
+            return
```
Comment (Faholan) [inline bot/cogs/moderation.py:141]: Why make it two lines ?

File: bot/cogs/moderation.py [hunk 33]
```diff
@@ -181,12 +198,142 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
         await asyncio.sleep(2.5)
         await message.delete()
 
+    @command()
+    @has_permissions(manage_messages=True)
+    async def shift(self, ctx, count: int, target: TextChannel, copy: bool = False) -> None:
+        """Copy or Move specified messages amount to specified channel"""
+        if not (5 <= count <= 150):
+            await ctx.send("Amount of messages shifted must be greater than 0 and smaller than 150")
+            return
```
Comment (Faholan) [inline bot/cogs/moderation.py:213]: You can make that only one line (that's the behaviour recommended on the `discord.py` server)
That goes for every repetition of this pattern

File: bot/cogs/moderation.py [hunk 34]
```diff
@@ -181,12 +198,142 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
                 """
             ),
-            color=discord.Color.orange(),
+            color=Color.orange(),
         )
         message = await ctx.send(ctx.author.mention, embed=embed)
         await asyncio.sleep(2.5)
         await message.delete()
 
+    @command()
```
Comment (Faholan) [inline bot/cogs/moderation.py:207]:
```suggestion
    @command(aliases=["copy"])
```

File: bot/cogs/moderation.py [hunk 35]
```diff
@@ -181,12 +198,142 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+    @has_permissions(administrator=True)
+    async def dm(self, ctx: Context, members: Greedy[t.Union[Member, Role]], *, text: str = None) -> None:
+        """Dm a List of Specified User from Your Guild."""
+        embed_data = self.embeds_cog.embeds[ctx.author]
+
+        if embed_data.embed.description is None and embed_data.embed.title is None:
+            await ctx.send("Please create a embed using our embed handler to send it.")
+            return
+
+        if text is not None:
```
Comment (Faholan) [inline bot/cogs/moderation.py:241]:
```suggestion
        if text:
```

File: bot/cogs/moderation.py [hunk 36]
```diff
@@ -181,12 +198,142 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+        for member in members:
+            if isinstance(member, Role):
+                for mem in member.members:
+                    await mem.send(embed=embed_data.embed)
+            else:
+                await member.send(embed=embed_data.embed)
+
+    @command()
+    @has_permissions(administrator=True)
+    async def dmall(self, ctx: Context, *, text: str = None) -> None:
```
Comment (Faholan) [inline bot/cogs/moderation.py:255]: Users should be able to opt out

File: bot/cogs/moderation.py [hunk 37]
```diff
@@ -181,12 +198,142 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+
+        await target.send(f'Message source : {ctx.channel.mention}.')
+
+        for embed in reversed(messages):
+            await target.send(embed=embed)
+            asyncio.sleep(0.5)
+
+    @command()
+    @has_permissions(administrator=True)
+    async def dm(self, ctx: Context, members: Greedy[t.Union[Member, Role]], *, text: str = None) -> None:
```
Comment (Faholan) [inline bot/cogs/moderation.py:233]: users should be able to opt out

File: bot/cogs/moderation.py [hunk 38]
```diff
@@ -181,12 +198,142 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+            if channel.permissions_for(ctx.author).manage_channels:
+                await channel.set_permissions(
+                    channel.guild.default_role,
+                    send_messages=False,
+                    reason=f"Reason: {reason} | Requested by {ctx.author}."
+                )
+                channel_count += 1
+            else:
+                continue
+            await channel.send("🔒 Locked down this channel.")
```
Comment (Faholan) [inline bot/cogs/moderation.py:290]:
```suggestion
            if channel.permissions_for(ctx.author).manage_channels:
                await channel.set_permissions(
                    channel.guild.default_role,
                    send_messages=False,
                    reason=f"Reason: {reason} | Requested by {ctx.author}."
                )
                channel_count += 1
                await channel.send("🔒 Locked down this channel.")   
```

File: bot/cogs/moderation.py [hunk 39]
```diff
@@ -181,12 +198,142 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+        for channel in channels:
+            if channel.permissions_for(ctx.author).manage_channels:
+                await channel.set_permissions(
+                    channel.guild.default_role,
+                    send_messages=None,
+                    reason=f"Reason: {reason} | Requested by {ctx.author}."
+                )
+                channel_count += 1
+            else:
+                continue
```
Comment (Faholan) [inline bot/cogs/moderation.py:311]:
```suggestion
```
That's completely useless

File: bot/cogs/moderation.py [hunk 40]
```diff
@@ -181,12 +198,142 @@ async def clear(self, ctx: Context, amount: int) -> None:
...
+        if channels is None:
+            channels = [ctx.channel]
+
+        channel_count = 0
+        for channel in channels:
+            if channel.permissions_for(ctx.author).manage_channels:
+                await channel.edit(reason=f"Reason: {reason} | Requested by {ctx.author}.", slowmode_delay=seconds)
+                channel_count += 1
+            else:
+                continue
```
Comment (Faholan) [inline bot/cogs/moderation.py:331]:
```suggestion
```
Same

File: bot/cogs/sudo.py [hunk 41]
```diff
@@ -40,77 +45,64 @@ def get_uptime(self) -> str:
         minutes, seconds = divmod(rem, 60)
         days, hours = divmod(hours, 24)
         if days:
-            return f"{days} days, {hours} hr, {minutes} mins, and {seconds} secs"
+            str = f"{days} days, {hours} hr, {minutes} mins, and {seconds} secs"
         else:
-            return f"{hours} hr, {minutes} mins, and {seconds} secs"
+            str = f"{hours} hr, {minutes} mins, and {seconds} secs"
```
Comment (Faholan) [inline bot/cogs/sudo.py:50]: NO !
`str` is a reserved word in python. NEVER USE IT FOR A VARIABLE NAME

File: bot/cogs/sudo.py [hunk 42]
```diff
@@ -40,77 +45,64 @@ def get_uptime(self) -> str:
         minutes, seconds = divmod(rem, 60)
         days, hours = divmod(hours, 24)
         if days:
-            return f"{days} days, {hours} hr, {minutes} mins, and {seconds} secs"
+            str = f"{days} days, {hours} hr, {minutes} mins, and {seconds} secs"
```
Comment (Faholan) [inline bot/cogs/sudo.py:48]: NO !
`str` is a reserved word in python. NEVER USE IT FOR A VARIABLE NAME

**Review by ActuallyDaneel — APPROVED** [review]
looks good to me

File: bot/cogs/sudo.py [hunk 43]
```diff
@@ -40,77 +45,64 @@ def get_uptime(self) -> str:
...
-        else:
-            embed = Embed(description="This is an owner-only command, you cannot invoke this.", color=Color.red())
-            await ctx.send(embed=embed)
+        return str
 
     @group(hidden=True)
-    @check(is_owner)
     async def sudo(self, ctx: Context) -> None:
         """Administrative information."""
         pass
```
Comment (Faholan) [inline bot/cogs/sudo.py:57]:
```suggestion
```
You don't need it (the docstring is enough)
You can also do this :
```suggestion
        await ctx.send_help("tag global")
```

File: bot/cogs/sudo.py [hunk 44]
```diff
@@ -191,66 +182,13 @@ async def stats(self, ctx: Context) -> None:
...
-        guilds_list = ''
-        for guild in guild_list[start:end]:
-            guilds_list += f'**{guild.name}** ({guild.id})\n**Joined:** {humanize.naturaltime(guild.get_member(self.bot.user.id).joined_at)}\n>'
-            guilds_list += "=====================================\n"
-
-        embed = Embed(color=Color.greyple(), title="Total Guilds", description=guilds_list)
-        embed.set_footer(text=f"Currently showing: {page} out of {pages}")
-
-        await ctx.send(embed=embed)
+    async def cog_check(self, ctx: Context) -> t.Optional[bool]:
```
Comment (Faholan) [inline bot/cogs/sudo.py:185]:
```suggestion
    async def cog_check(self, ctx: Context) -> bool:
```
Why `Optional` ?

File: bot/utils/checks.py [hunk 45]
```diff
@@ -1,8 +1,8 @@
+from bot.config import devs
+
 from discord import Member
 from discord.ext.commands import Context, NoPrivateMessage
```
Comment (Faholan) [inline bot/utils/checks.py:5]:
```suggestion
from discord import Member
from discord.ext.commands import Context, NoPrivateMessage

from bot.config import devs

```
Wrong import order : local import should be located after third party import

File: bot/cogs/sudo.py [hunk 46]
```diff
@@ -1,21 +1,26 @@
...
+    ActivityType,
+    Color,
+    Embed,
+    Game,
+    InvalidArgument,
+    Status
+)
+from discord import __version__ as discord_version
+from discord.ext.commands import Cog, Context, group
+
```
Comment (Faholan) [inline bot/cogs/sudo.py:23]:
```suggestion
from discord import (
    Activity,
    ActivityType,
    Color,
    Embed,
    Game,
    InvalidArgument,
    Status
)
from discord import __version__ as discord_version
from discord.ext.commands import Cog, Context, group

from bot.config import devs
from bot.core.bot import Bot

```
Wrong import order

File: bot/cogs/moderation.py [hunk 47]
```diff
@@ -2,26 +2,38 @@
...
+from discord import Color, Embed, Member, NotFound, Role, TextChannel, HTTPException
+from discord.errors import Forbidden
+from discord.ext.commands import (
+    Cog,
+    Context,
+    Greedy,
+    NoPrivateMessage,
+    command,
+    has_permissions
+)
```
Comment (Faholan) [inline bot/cogs/moderation.py:22]:
```suggestion
from discord import Color, Embed, Member, NotFound, Role, TextChannel, HTTPException
from discord.errors import Forbidden
from discord.ext.commands import (
    Cog,
    Context,
    Greedy,
    NoPrivateMessage,
    command,
    has_permissions
)

from bot.core.bot import Bot
from bot.core.converters import ActionReason, ProcessedMember
from bot.core.decorators import follow_roles
from bot.utils.formats import Plural
```
Wrong import order

Comment (Faholan) [reply bot/cogs/moderation.py:22]: Change it in every file please (typically bot/cogs/commands.py)

**Review by Faholan — CHANGES_REQUESTED** [review]
A few changes are necessary
Biggest of all is to fix the import order in every single file
